File: jvm/selfie-lib/src/commonMain/kotlin/com/diffplug/selfie/guts/SourceFile.kt
/*
 * Copyright (C) 2023-2025 DiffPlug
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.diffplug.selfie.guts

import com.diffplug.selfie.efficientReplace

private const val TRIPLE_QUOTE = "\"\"\""

/**
 * @param filename The filename (not full path, but the extension is used for language-specific
 *   parsing).
 * @param content The exact content of the file, unix or windows newlines will be preserved
 */
class SourceFile(filename: String, content: String, val language: Language) {
  private val unixNewlines = content.indexOf('\r') == -1
  private var contentSlice = Slice(content.efficientReplace("\r\n", "\n"))
  private val escapeLeadingWhitespace =
      EscapeLeadingWhitespace.appropriateFor(contentSlice.toString())

  /**
   * Returns the content of the file, possibly modified by
   * [ToBeLiteral.setLiteralAndGetNewlineDelta].
   */
  val asString: String
    get() = contentSlice.toString().let { if (unixNewlines) it else it.replace("\n", "\r\n") }

  /**
   * Represents a section of the sourcecode which is a `.toBe(LITERAL)` call. It might also be
   * `.toBe_TODO()` or ` toBe LITERAL` (infix notation).
   */
  inner class ToBeLiteral
  internal constructor(
      private val dotFunOpenParen: String,
      internal val functionCallPlusArg: Slice,
      internal val arg: String,
  ) {
    /**
     * Modifies the parent [SourceFile] to set the value within the `toBe` call, and returns the net
     * change in newline count.
     */
    fun <T : Any> setLiteralAndGetNewlineDelta(literalValue: LiteralValue<T>): Int {
      val encoded =
          literalValue.format.encode(literalValue.actual, language, escapeLeadingWhitespace)
      val roundTripped = literalValue.format.parse(encoded, language) // sanity check
      if (roundTripped != literalValue.actual) {
        throw Error(
            "There is an error in " +
                literalValue.format::class.simpleName +
                ", the following value isn't round tripping.\n" +
                "Please this error and the data below at https://github.com/diffplug/selfie/issues/new\n" +
                "```\n" +
                "ORIGINAL\n" +
                literalValue.actual +
                "\n" +
                "ROUNDTRIPPED\n" +
                roundTripped +
                "\n" +
                "ENCODED ORIGINAL\n" +
                encoded +
                "\n" +
                "```\n")
      }
      val existingNewlines = functionCallPlusArg.count { it == '\n' }
      val newNewlines = encoded.count { it == '\n' }
      contentSlice = Slice(functionCallPlusArg.replaceSelfWith("${dotFunOpenParen}${encoded})"))
      return newNewlines - existingNewlines
    }

    /**
     * Parses the current value of the value within `.toBe()`. This method should not be called on
     * `toBe_TODO()`.
     */
    fun <T : Any> parseLiteral(literalFormat: LiteralFormat<T>): T {
      return literalFormat.parse(arg, language)
    }
  }
  fun removeSelfieOnceComments() {
    contentSlice = Slice(RemoveSelfieOnceComment.removeSelfieComment(contentSlice.toString()))
  }
  private fun findOnLine(toFind: String, lineOneIndexed: Int): Slice {
    val lineContent = contentSlice.unixLine(lineOneIndexed)
    val idx = lineContent.indexOf(toFind)
    if (idx == -1) {
      throw AssertionError(
          "Expected to find `$toFind` on line $lineOneIndexed, but there was only `${lineContent}`")
    }
    return lineContent.subSequence(idx, idx + toFind.length)
  }
  fun replaceOnLine(lineOneIndexed: Int, find: String, replace: String) {
    check(find.indexOf('\n') == -1)
    check(replace.indexOf('\n') == -1)
    val slice = findOnLine(find, lineOneIndexed)
    contentSlice = Slice(slice.replaceSelfWith(replace))
  }
  fun parseToBeLike(lineOneIndexed: Int): ToBeLiteral {
    val lineContent = contentSlice.unixLine(lineOneIndexed)
    val dotFunOpenParen =
        TO_BE_LIKES.mapNotNull {
              val idx = lineContent.indexOf(it)
              if (idx == -1) null else idx to it
            }
            .minByOrNull { it.first }
            ?.second
            ?: throw AssertionError(
                "Expected to find inline assertion on line $lineOneIndexed, but there was only `${lineContent}`")
    val dotFunctionCallInPlace = lineContent.indexOf(dotFunOpenParen)
    val dotFunctionCall = dotFunctionCallInPlace + lineContent.startIndex
    val argStart = dotFunctionCall + dotFunOpenParen.length
    if (contentSlice.length == argStart) {
      throw AssertionError(
          "Appears to be an unclosed function call `$dotFunOpenParen)` on line $lineOneIndexed")
    }

    var commaDelimitedNewlines: MutableList<Slice>? = null
    val arg = argSlice(argStart, dotFunOpenParen, lineOneIndexed)
    var endParen = arg.endIndex
    while (contentSlice[endParen] != ')') {
      val nextChar = contentSlice[endParen]
      if (nextChar == ',') {
        val nextArg = argSlice(endParen + 1, dotFunOpenParen, lineOneIndexed)
        if (commaDelimitedNewlines == null) {
          commaDelimitedNewlines = mutableListOf(arg, nextArg)
        } else {
          commaDelimitedNewlines.add(nextArg)
        }
        endParen = nextArg.endIndex - 1
      } else if (!nextChar.isWhitespace()) {
        throw AssertionError(
            "Non-primitive literal in `$dotFunOpenParen)` starting at line $lineOneIndexed: error for character `${contentSlice[endParen]}` on line ${contentSlice.baseLineAtOffset(endParen)}")
      }
      ++endParen
      if (endParen == contentSlice.length) {
        throw AssertionError(
            "Appears to be an unclosed function call `$dotFunOpenParen)` starting at line $lineOneIndexed")
      }
    }
    val fullArg =
        if (commaDelimitedNewlines == null) arg.toString()
        else
            commaDelimitedNewlines.joinToString("\\n", "\"", "\"") {
              check(it.startsWith("\"") && it.endsWith("\"")) {
                "Expected string literal to start with a single quote on line ${
                    it.baseLineAtOffset(
                        0
                    )
                }"
              }
              it.subSequence(1, it.length - 1)
            }
    return ToBeLiteral(
        dotFunOpenParen.replace("_TODO", ""),
        contentSlice.subSequence(dotFunctionCall, endParen + 1),
        fullArg)
  }

  companion object {
    internal inline fun commaDelimitedParseCleanup(str: String): String =
        str.efficientReplace("\",\n\"", "\\n")
  }
  private fun argSlice(
      argStartInitial: Int,
      dotFunOpenParen: String,
      lineOneIndexed: Int,
  ): Slice {
    var argStart = argStartInitial
    while (contentSlice[argStart].isWhitespace()) {
      ++argStart
      if (contentSlice.length == argStart) {
        throw AssertionError(
            "Appears to be an unclosed function call `$dotFunOpenParen)` on line $lineOneIndexed")
      }
    }
    // argStart is now the first non-whitespace character after the opening paren
    var endArg: Int
    if (contentSlice[argStart] == '"') {
      if (contentSlice.subSequence(argStart, contentSlice.length).startsWith(TRIPLE_QUOTE)) {
        endArg = contentSlice.indexOf(TRIPLE_QUOTE, argStart + TRIPLE_QUOTE.length)
        if (endArg == -1) {
          throw AssertionError(
              "Appears to be an unclosed multiline string literal `${TRIPLE_QUOTE}` on line $lineOneIndexed")
        } else {
          endArg += TRIPLE_QUOTE.length
        }
      } else {
        endArg = argStart + 1
        while (contentSlice[endArg] != '"' || contentSlice[endArg - 1] == '\\') {
          ++endArg
          if (endArg == contentSlice.length) {
            throw AssertionError(
                "Appears to be an unclosed string literal `\"` on line $lineOneIndexed")
          }
        }
        endArg += 1
      }
    } else {
      endArg = argStart
      while (!contentSlice[endArg].isWhitespace()) {
        if (contentSlice[endArg] == ')') {
          break
        }
        ++endArg
        if (endArg == contentSlice.length) {
          throw AssertionError("Appears to be an unclosed numeric literal on line $lineOneIndexed")
        }
      }
    }
    return contentSlice.subSequence(argStart, endArg)
  }
}
private val TO_BE_LIKES = listOf(".toBe(", ".toBe_TODO(", ".toBeBase64(", ".toBeBase64_TODO(")

File: python/selfie-lib/selfie_lib/SourceFile.py
from typing import Any

from .EscapeLeadingWhitespace import EscapeLeadingWhitespace
from .Literals import Language, LiteralFormat, LiteralRepr, LiteralValue
from .Slice import Slice


class SourceFile:
    TRIPLE_QUOTE = '"""'

    def __init__(self, filename: str, content: str) -> None:
        self.__unix_newlines: bool = "\r" not in content
        self._content_slice: Slice = Slice(content.replace("\r\n", "\n"))
        self.__language: Language = Language.from_filename(filename)
        self.__escape_leading_whitespace = EscapeLeadingWhitespace.appropriate_for(
            self._content_slice.__str__()
        )

    def remove_selfie_once_comments(self):
        # Split content into lines
        lines = self._content_slice.__str__().split("\n")

        # Create a new list of lines, excluding lines containing '# selfieonce' or '#selfieonce'
        new_lines = []
        for line in lines:
            # Check for both variations of the comment
            if "# selfieonce" in line:
                cleaned_line = line.split("# selfieonce")[0].strip()
            elif "#selfieonce" in line:
                cleaned_line = line.split("#selfieonce")[0].strip()
            else:
                new_lines.append(line)
                continue

            # If the line has code before the comment, keep the code part
            if cleaned_line:
                new_lines.append(cleaned_line)

        # Recombine the lines into a single string
        new_content = "\n".join(new_lines)

        # Update the content slice with new content
        self._content_slice = Slice(new_content)

        if not self.__unix_newlines:
            self._content_slice = Slice(new_content.replace("\n", "\r\n"))

    @property
    def as_string(self) -> str:
        return (
            self._content_slice.__str__()
            if self.__unix_newlines
            else self._content_slice.__str__().replace("\n", "\r\n")
        )

    class ToBeLiteral:
        def __init__(
            self,
            parent: "SourceFile",
            dot_fun_open_paren: str,
            function_call_plus_arg: Slice,
            arg: Slice,
            language: Language,
            escape_leading_whitespace: EscapeLeadingWhitespace,
        ) -> None:
            self.__parent = parent
            self.__dot_fun_open_paren = dot_fun_open_paren
            self.__function_call_plus_arg = function_call_plus_arg
            self.__arg = arg
            self.__language = language
            self.__escape_leading_whitespace = escape_leading_whitespace

        def _get_function_call_plus_arg(self):
            return self.__function_call_plus_arg

        def _get_arg(self):
            return self.__arg

        def set_literal_and_get_newline_delta(self, literal_value: LiteralValue) -> int:
            encoded = literal_value.format.encode(
                literal_value.actual,
                self.__language,
                self.__escape_leading_whitespace,
            )
            if not isinstance(literal_value.format, LiteralRepr):
                # we don't roundtrip LiteralRepr because `eval` is dangerous
                round_tripped = literal_value.format.parse(encoded, self.__language)
                if round_tripped != literal_value.actual:
                    raise ValueError(
                        f"There is an error in {literal_value.format.__class__.__name__}, "
                        "the following value isn't round tripping.\n"
                        f"Please report this error and the data below at "
                        "https://github.com/diffplug/selfie/issues/new\n"
                        f"```\n"
                        f"ORIGINAL\n{literal_value.actual}\n"
                        f"ROUNDTRIPPED\n{round_tripped}\n"
                        f"ENCODED ORIGINAL\n{encoded}\n"
                        f"```\n"
                    )
            existing_newlines = self.__function_call_plus_arg.count("\n")
            new_newlines = encoded.count("\n")
            self.__parent._content_slice = Slice(  # noqa: SLF001
                self.__function_call_plus_arg.replaceSelfWith(
                    f"{self.__dot_fun_open_paren}{encoded})"
                )
            )

            return new_newlines - existing_newlines

        def parse_literal(self, literal_format: LiteralFormat) -> Any:
            return literal_format.parse(self.__arg.__str__(), self.__language)

    def find_on_line(self, to_find: str, line_one_indexed: int) -> Slice:
        line_content = self._content_slice.unixLine(line_one_indexed)
        idx = line_content.indexOf(to_find)
        if idx == -1:
            raise AssertionError(
                f"Expected to find `{to_find}` on line {line_one_indexed}, but there was only `{line_content}`"
            )
        return line_content.subSequence(idx, idx + len(to_find))

    def replace_on_line(self, line_one_indexed: int, find: str, replace: str) -> None:
        assert "\n" not in find
        assert "\n" not in replace

        found = self.find_on_line(find, line_one_indexed)
        self._content_slice = Slice(found.replaceSelfWith(replace))

    def parse_to_be_like(self, line_one_indexed: int) -> ToBeLiteral:
        line_content = self._content_slice.unixLine(line_one_indexed)
        dot_fun_open_paren = None

        for to_be_like in TO_BE_LIKES:
            idx = line_content.indexOf(to_be_like)
            if idx != -1:
                dot_fun_open_paren = to_be_like
                break
        if dot_fun_open_paren is None:
            raise AssertionError(
                f"Expected to find inline assertion on line {line_one_indexed}, but there was only `{line_content}`"
            )

        dot_function_call_in_place = line_content.indexOf(dot_fun_open_paren)
        dot_function_call = dot_function_call_in_place + line_content.startIndex
        arg_start = dot_function_call + len(dot_fun_open_paren)

        if self._content_slice.__len__() == arg_start:
            raise AssertionError(
                f"Appears to be an unclosed function call `{dot_fun_open_paren}` "
                f"on line {line_one_indexed}"
            )
        while self._content_slice[arg_start].isspace():
            arg_start += 1
            if self._content_slice.__len__() == arg_start:
                raise AssertionError(
                    f"Appears to be an unclosed function call `{dot_fun_open_paren}` "
                    f"on line {line_one_indexed}"
                )

        if self._content_slice[arg_start] == '"':
            (end_paren, end_arg) = self._parse_string(
                line_one_indexed, arg_start, dot_fun_open_paren
            )
        else:
            (end_paren, end_arg) = self._parse_code(
                line_one_indexed, arg_start, dot_fun_open_paren
            )
        return self.ToBeLiteral(
            self,
            dot_fun_open_paren.replace("_TODO", ""),
            self._content_slice.subSequence(dot_function_call, end_paren + 1),
            self._content_slice.subSequence(arg_start, end_arg),
            self.__language,
            self.__escape_leading_whitespace,
        )

    def _parse_code(
        self,
        line_one_indexed: int,
        arg_start: int,
        dot_fun_open_paren: str,
    ):
        # Initialize variables
        parenthesis_count = 1
        string_delimiter = None

        # Iterate through the characters starting from the given index
        for i in range(arg_start, len(self._content_slice)):
            char = self._content_slice[i]

            # Check if we are entering or leaving a string
            if char in ["'", '"'] and self._content_slice[i - 1] != "\\":
                if not string_delimiter:
                    string_delimiter = char
                elif char == string_delimiter:
                    string_delimiter = None

            # Skip characters inside strings
            if string_delimiter:
                continue

            # Count parentheses
            if char == "(":
                parenthesis_count += 1
            elif char == ")":
                parenthesis_count -= 1

            # If all parentheses are closed, return the current index
            if parenthesis_count == 0:
                end_paren = i
                end_arg = i
                return (end_paren, end_arg)
        # else ...
        raise AssertionError(
            f"Appears to be an unclosed function call `{dot_fun_open_paren}` "
            f"starting at line {line_one_indexed}"
        )

    def _parse_string(
        self,
        line_one_indexed: int,
        arg_start: int,
        dot_fun_open_paren: str,
    ):
        if self._content_slice.subSequence(
            arg_start, len(self._content_slice)
        ).starts_with(self.TRIPLE_QUOTE):
            end_arg = self._content_slice.indexOf(
                self.TRIPLE_QUOTE, arg_start + len(self.TRIPLE_QUOTE)
            )
            if end_arg == -1:
                raise AssertionError(
                    f"Appears to be an unclosed multiline string literal `{self.TRIPLE_QUOTE}` "
                    f"on line {line_one_indexed}"
                )
            else:
                end_arg += len(self.TRIPLE_QUOTE)
                end_paren = end_arg
        else:
            end_arg = arg_start + 1
            while (
                self._content_slice[end_arg] != '"'
                or self._content_slice[end_arg - 1] == "\\"
            ):
                end_arg += 1
                if end_arg == self._content_slice.__len__():
                    raise AssertionError(
                        f'Appears to be an unclosed string literal `"` '
                        f"on line {line_one_indexed}"
                    )
            end_arg += 1
            end_paren = end_arg
        while self._content_slice[end_paren] != ")":
            if not self._content_slice[end_paren].isspace():
                raise AssertionError(
                    f"Non-primitive literal in `{dot_fun_open_paren}` starting at "
                    f"line {line_one_indexed}: error for character "
                    f"`{self._content_slice[end_paren]}` on line "
                    f"{self._content_slice.baseLineAtOffset(end_paren)}"
                )
            end_paren += 1
            if end_paren == self._content_slice.__len__():
                raise AssertionError(
                    f"Appears to be an unclosed function call `{dot_fun_open_paren}` "
                    f"starting at line {line_one_indexed}"
                )
        return (end_paren, end_arg)


TO_BE_LIKES = [
    ".to_be(",
    ".to_be_TODO(",
    ".to_be_base64(",
    ".to_be_base64_TODO(",
    ".to_be_TODO(",
]
